### PR TITLE
[DataGrid] Fix onPageChange firing too often

### LINKED ISSUE #816: [DataGrid] Events firing multiple times and also while table initially renders
- [x] The issue is present in the latest release.
- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

- Current behaviour introduced a bug that fires onPageChange event multiple times
- Even when the XGrid renders the event fires thrice (one client event and twice server)

## Expected Behavior 🤔

- Not to fire events on table render
- Event should fire once on its use case

## Steps to Reproduce 🕹

DataGrid: https://codesandbox.io/s/material-demo-forked-9jzwr?file=/package.json
XGrid: https://codesandbox.io/s/material-demo-forked-8dylq?file=/demo.tsx

Steps:

1. Reload the template, you can see onPageChange event firing thrice.
2. Go on to the next page, event fires twice.
3. Go back to the previous page, same issue.

## Context 🔦

- Due to multiple event firing app is re-rendering eventually causing performance issues.

## Your Environment 🌎

<details>
  <summary>`npx @material-ui/envinfo`</summary>
  
```
 System:
    OS: Linux 5.4 Ubuntu 20.04.1 LTS (Focal Fossa)
  Binaries:
    Node: 12.20.1 - /usr/bin/node
    Yarn: 1.22.5 - /usr/bin/yarn
    npm: 6.14.10 - /usr/bin/npm
  Browsers:
    Chrome: 87.0.4280.88
    Firefox: Not Found
  npmPackages:
    @emotion/styled:  10.0.27 
    @material-ui/core: 4.11.0 => 4.11.0 
    @material-ui/icons: 4.9.1 => 4.9.1 
    @material-ui/lab: 4.0.0-alpha.56 => 4.0.0-alpha.56 
    @material-ui/pickers:  3.2.10 
    @material-ui/styles:  4.11.1 
    @material-ui/system:  4.9.14 
    @material-ui/types:  5.1.0 
    @material-ui/utils:  4.10.2 
    @material-ui/x-grid: 4.0.0-alpha.14 => 4.0.0-alpha.14 
    @material-ui/x-license:  4.0.0-alpha.14 
    @types/react: 16.9.2 => 16.9.2 
    react: 16.9.0 => 16.9.0 
    react-dom: 16.9.0 => 16.9.0 
    typescript: 4.0.2 => 4.0.2 

```
</details>

### PR DESCRIPTION
**Pull request by dtassone**
Fix #816, onPageChange firing multiple times

### CODE REVIEW

File: packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx [hunk 1]
```diff
@@ -67,6 +68,40 @@ describe('<DataGrid /> - Pagination', () => {
...
 
+    it('should trigger onPageChange once if page prop is set', ()=> {
+      const onPageChange = spy();
+
+      render(
+        <div style={{ width: 300, height: 300 }}>
+          <DataGrid {...baselineProps} page={1} onPageChange={onPageChange} pageSize={1} />
+        </div>,
+      );
+      expect(onPageChange.calledOnce).to.be.true;
```
Comment (oliviertassinari) [inline packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx:79]: ?

```suggestion
      expect(onPageChange.callCount).to.equal(1);
```

File: packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx [hunk 2]
```diff
@@ -67,6 +68,40 @@ describe('<DataGrid /> - Pagination', () => {
...
+    it('should trigger onPageChange when clicking on next page', ()=> {
+      const onPageChange = spy();
+
+      render(
+        <div style={{ width: 300, height: 300 }}>
+          <DataGrid {...baselineProps} onPageChange={onPageChange} pageSize={1} />
+        </div>,
+      );
+      fireEvent.click(screen.getByRole('button', { name: /next page/i }));
+      expect(onPageChange.calledOnce).to.be.true;
```
Comment (oliviertassinari) [inline packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx:91]:
```suggestion
      expect(onPageChange.callCount).to.equal(1);
```

File: packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx [hunk 3]
```diff
@@ -67,6 +68,40 @@ describe('<DataGrid /> - Pagination', () => {
...
+
+    it('should not trigger onPageChange on initialisation and rendering of the first and default page', ()=> {
+      const onPageChange = spy();
+
+      render(
+        <div style={{ width: 300, height: 300 }}>
+          <DataGrid {...baselineProps} onPageChange={onPageChange} pageSize={1} />
+        </div>,
+      );
+      expect(onPageChange.calledOnce).to.be.false;
```
Comment (oliviertassinari) [inline packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx:102]:
```suggestion
      expect(onPageChange.callCount).to.equal(0);
```

File: packages/grid/_modules_/grid/hooks/features/pagination/usePagination.ts [hunk 4]
```diff
@@ -102,12 +102,6 @@ export const usePagination = (apiRef: ApiRef): void => {
     dispatch(setRowCountActionCreator({ totalRowCount: visibleRowCount }));
   }, [apiRef, dispatch, visibleRowCount]);
 
-  React.useEffect(() => {
```
Comment (DanailH) [inline packages/grid/_modules_/grid/hooks/features/pagination/usePagination.ts:105]: I tried to quickly fix it by doing that but if I remember correctly removing this messes up the initial render

**Comment by oliviertassinari** [comment]
Capture d’écran 2021-01-11 à 17 35 28

@dtassone Regarding the format of the pull request, I would recommend the following:

- We link the issue in the PR description so:
  1. the issue is closed automatically once this PR is merged: https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
  2. the PR is referenced in the issue, right now, there is nothing:

Capture d’écran 2021-01-11 à 17 37 33

- We polish the title of the pull request to be usable in the changelog, maybe something like this:
  `[DataGrid] Fix onPageChange firing too often`

File: packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx (hunk 1 above)
Comment (dtassone) [reply packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx:79]: Well it does not mean it's the same page, what if the user changes the page and set it back to the first one?

File: packages/grid/_modules_/grid/hooks/features/pagination/usePagination.ts (hunk 4 above)
Comment (dtassone) [reply packages/grid/_modules_/grid/hooks/features/pagination/usePagination.ts:105]: Any idea where I can see the issue?

File: packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx (hunk 1 above)
Comment (oliviertassinari) [reply packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx]: > what if the user changes the page and set it back to the first one?

The developer has provided the page prop, the only way the user can change the page is if the developer change the page prop. So in your case, the page doesn't change.

File: packages/grid/_modules_/grid/hooks/features/pagination/usePagination.ts (hunk 4 above)
Comment (dtassone) [reply packages/grid/_modules_/grid/hooks/features/pagination/usePagination.ts:105]: Found it :)

File: packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx [hunk 5]
```diff
@@ -67,6 +68,76 @@ describe('<DataGrid /> - Pagination', () => {
...
+    it('should trigger onPageChange on initialisation and rendering of the first and default page if Server mode', () => {
+      const onPageChange = spy();
+
+      render(
+        <div style={{ width: 300, height: 300 }}>
+          <DataGrid
+            {...baselineProps}
+            onPageChange={onPageChange}
+            pageSize={1}
+            paginationMode={'server'}
```
Comment (oliviertassinari) [inline packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx:134]:
```suggestion
            paginationMode="server"
```

File: packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx [hunk 6]
```diff
@@ -67,6 +68,76 @@ describe('<DataGrid /> - Pagination', () => {
...
+      const onPageChange = spy();
+
+      render(
+        <div style={{ width: 300, height: 300 }}>
+          <DataGrid {...baselineProps} onPageChange={onPageChange} pageSize={1} />
+        </div>,
+      );
+      expect(onPageChange.callCount).to.equal(0);
+    });
+    it('should trigger onPageChange on initialisation and rendering of the first and default page if Server mode', () => {
```
Comment (oliviertassinari) [inline packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx:125]:
```suggestion

    it('should trigger onPageChange on initialisation and rendering of the first and default page if Server mode', () => {
```

File: packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx [hunk 7]
```diff
@@ -67,6 +68,76 @@ describe('<DataGrid /> - Pagination', () => {
...
+    it('should trigger onPageChange when clicking on next page in Server mode', () => {
+      const onPageChange = spy();
+
+      render(
+        <div style={{ width: 300, height: 300 }}>
+          <DataGrid
+            {...baselineProps}
+            onPageChange={onPageChange}
+            pageSize={1}
+            paginationMode={'server'}
```
Comment (oliviertassinari) [inline packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx:103]:
```suggestion
            paginationMode="server"
```

File: packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx [hunk 8]
```diff
@@ -67,6 +68,76 @@ describe('<DataGrid /> - Pagination', () => {
...
 
+    it('should trigger onPageChange once if page prop is set', () => {
+      const onPageChange = spy();
+
+      render(
+        <div style={{ width: 300, height: 300 }}>
+          <DataGrid {...baselineProps} page={1} onPageChange={onPageChange} pageSize={1} />
+        </div>,
+      );
+      expect(onPageChange.callCount).to.equal(1);
```
Comment (oliviertassinari) [inline packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx:79]:
```suggestion
      expect(onPageChange.callCount).to.equal(0);
```

File: packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx [hunk 9]
```diff
@@ -67,6 +68,76 @@ describe('<DataGrid /> - Pagination', () => {
...
+            paginationMode={'server'}
+          />
+        </div>,
+      );
+      // First time on init
+      expect(onPageChange.callCount).to.equal(1);
+      fireEvent.click(screen.getByRole('button', { name: /next page/i }));
+      expect(onPageChange.callCount).to.equal(2);
+      fireEvent.click(screen.getByRole('button', { name: /next page/i }));
+      expect(onPageChange.callCount).to.equal(3);
```
Comment (oliviertassinari) [inline packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx:112]: The default page is 1, it shouldn't fire for it

```suggestion
      expect(onPageChange.callCount).to.equal(0);
      fireEvent.click(screen.getByRole('button', { name: /next page/i }));
      expect(onPageChange.callCount).to.equal(1);
      fireEvent.click(screen.getByRole('button', { name: /next page/i }));
      expect(onPageChange.callCount).to.equal(2);
```

**Review by oliviertassinari — COMMENTED** [review]
I have commented on my understanding of how it should behave. The mental model I have always seen for the change event is: "does somebody try to change the existing state". For instance, when doing server-side rendering, the default page is 1, it shouldn't fire.

File: packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx (hunk 8 above)
Comment (dtassone) [reply packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx:79]: > I have commented on my understanding of how it should behave. The mental model I have always seen for the change event is: "does something try to change the current state?". For instance, when doing server-side rendering, the default page is already 1, I don't see why it should fire.

Well, we can argue that when the grid initially loads, then it goes from page 0 to page 1. By firing this event on page 1 then developers would just load page 1 as any page using their onPageChange event, so they don't need to initialize the first page separately and use onPageChange for page 2.

File: packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx (hunk 9 above)
Comment (dtassone) [reply packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx:112]: I think it should fire to easily load the content of the first page like any other page

Comment (dtassone) [reply packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx:112]: This avoids to pass the content of the first page in the rows prop. Page 1 is treated like any other page

Comment (DanailH) [reply packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx:112]: I agree with @oliviertassinari . The description in the bug that was reported was that the expectation is that the event should be fired initially, only after the page change, which I think the reporter understands as interacting with the pagination (from page 1 to page 2 for example).

File: packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx (hunk 8 above)
Comment (oliviertassinari) [reply packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx:79]: Do you have more details on page 0? What does this state mean? Is it valid? Why doesn't it alway start on page 1?

Comment (oliviertassinari) [reply packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx:79]: > so they don't need to initialize the first page separately and use onPageChange for page 2.

This logic doesn't work server-side.

**Review by DanailH — APPROVED** [review]

File: packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx (hunk 8 above)
Comment (oliviertassinari) [reply packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx:79]: On a related note, I think that we should consider changing the index for the default page. The Pagination component, which is meant for SEO starts at page 1. The TablePagination component which is meant for data grid use cases starts at page 0: https://material-ui.com/components/pagination/#table-pagination. It looks like DataGrid starts at page 1: https://material-ui.com/components/data-grid/pagination/#controlled-pagination. It's inconsistent.

Comment (dtassone) [reply packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx:79]: Can we create a separate issue for that pls

Comment (oliviertassinari) [reply packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx:79]: Sure, done

**Pull request merged**